generate/main.go

```diff
@@ -0,0 +1,205 @@
+// TODO: implement function that merges the paths of each Swagger spec into one file before running codegen
+var MergedClient = false
+
+// trimVersionTag : remove the "v" prefix from the version tag string
+func trimVersionTag(tag string) string {
+	if strings.HasPrefix(tag, "v") {
+		return strings.TrimPrefix(tag, "v")
+	}
+	return tag
+}
+
+func getDefaultVersion() string {
+	version := os.Getenv("VERSION")
+	if version == "" {
+		tag := getMostRecentTag()
+		trimmedVersion := trimVersionTag(tag)
+		version = trimmedVersion
+	}
+	return version
+}
+
+func getMostRecentTag() string {
+	tagsBytes, _ := get("https://api.github.com/repos/grafeas/grafeas/tags")
+	var tags []map[string]interface{}
+	json.Unmarshal(tagsBytes, &tags)
+	return tags[0]["name"].(string)
+}
+
+func swaggerCompatibility(jsonInput string) string {
+	re := regexp.MustCompile(`\/{.*(=.*)}`)
+	jsonOutput := re.ReplaceAllStringFunc(jsonInput, func(substringMatch string) string {
+		return strings.Split(substringMatch, "=")[0] + "}"
+	})
+	return jsonOutput
+}
+
+func get(url string) ([]byte, error) {
+	var responseData []byte
+	resp, err := http.Get(url)
+	if err != nil {
+		return responseData, err
+	}
+	defer resp.Body.Close()
+
+	responseData, err = ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return responseData, err
+	}
+	return responseData, err
+}
+
+func downloadCompatibleSwaggerSpec(url string, filename string) error {
+	if filename == "" {
+		filename = path.Base(url)
+	}
+	responseBytes, err := get(url)
+	if err != nil {
+		return err
+	}
+
+	swaggerString := swaggerCompatibility(string(responseBytes))
+
+	f, err := os.Create(filename)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	responseReader := bytes.NewReader([]byte(swaggerString))
+
+	_, err = io.Copy(f, responseReader)
+	return err
+}
+
+func downloadFile(url string, filename string) error {
+	if filename == "" {
+		filename = path.Base(url)
+	}
+	responseBytes, err := get(url)
+	if err != nil {
+		return err
+	}
+
+	f, err := os.Create(filename)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	responseReader := bytes.NewReader(responseBytes)
+
+	_, err = io.Copy(f, responseReader)
+	return err
+}
+
+type swaggerCodegenConfig struct {
+	language        string
+	inputSpec       string
+	outputDirectory string
+	configFile      string
+	jar             string
+}
+
+func stringOrDefault(currentValue string, defaultValue string) string {
+	if currentValue == "" {
+		return defaultValue
+	}
+	return currentValue
+}
+
+func (c *swaggerCodegenConfig) generate() {
+	c.language = stringOrDefault(c.language, "go")
+	c.jar = stringOrDefault(c.jar, "./swagger-codegen-cli.jar")
+	c.configFile = stringOrDefault(c.configFile, "./config.go.json")
+	c.outputDirectory = stringOrDefault(c.outputDirectory, ServerVersion)
+	args := []string{"-jar", c.jar, "generate", "-i", c.inputSpec, "-l", c.language, "-o", c.outputDirectory, "-c", c.configFile}
```

Comment:
Could use this logic instead of updating and committing the version in the `config.json` each time:
```go
args := []string{"-jar", c.jar, "generate", "-i", c.inputSpec, "-l", c.language, "-o", c.outputDirectory, "-c", c.configFile, fmt.Sprintf("-DpackageVersion=", ServerVersion)}
```
Or something like this to forego the file entirely
```go
args := []string{"-jar", c.jar, "generate", "-i", c.inputSpec, "-l", c.language, "-o", c.outputDirectory, "-DpackageName=grafeas", fmt.Sprintf("-DpackageVersion=", ServerVersion)}
```

Reply:
To clarify: what would be the difference in output in the 3 versions discussed here?

Reply:
Hopefully this helps a little bit

## Version 1 (as-is implementation)

This is essentially "feature parity" with the existing process. It relies on the `config.go.json` being updated by the user _prior_ to running the client generate.

* `config.go.json` provides both the `packageVersion` and the `packageName` in this example
* command line args remain "unmodified" from the existing implementation

```go
args := []string{"-jar", c.jar, "generate", "-i", c.inputSpec, "-l", c.language, "-o", c.outputDirectory, "-c", c.configFile}
```

Which would yield something like:

```shell
java -jar ./swagger-codegen-cli.jar generate -i ./grafeas.swagger.json -l go -o 0.1.0 -c ./config.go.json
```

## Version 2

This pulls in the "server version" that is being used to generate the client and overrides the `packageVersion` present in the `config.go.json` (or means, in essence, that you can omit this value from that file altogether). As a result, there is more of an assurance that the version is consistent, and could be more automation friendly.

The nice thing about this is that version can be passed in as one of the flags, or can be ingested via an environment variable. If neither are specified, this will grab the latest tag from the `grafeas/grafeas` repository and "release" that.

https://github.com/psprings/client-go/blob/swagger-compatibility/generate/main.go#L46-L54

* `config.go.json` provides only the `packageName`
* the command line args provide `packageVersion`

```go
args := []string{"-jar", c.jar, "generate", "-i", c.inputSpec, "-l", c.language, "-o", c.outputDirectory, "-c", c.configFile, fmt.Sprintf("-DpackageVersion=", ServerVersion)}
```

Which would yield something like:

```shell
java -jar ./swagger-codegen-cli.jar generate -i ./grafeas.swagger.json -l go -o 0.1.0 -c ./config.go.json -DpackageVersion=0.1.0
```

## Version 3

If **Version 2** was agreeable enough, it seems less meaningful to have a `config.go.json` file which _only provides_ a single value (`packageName`) so you might as well just leverage that as a command line arg.

* `config.go.json` provides nothing (and can be removed)
* the command line args provide both `packageName` and `packageVersion`

```go
args := []string{"-jar", c.jar, "generate", "-i", c.inputSpec, "-l", c.language, "-o", c.outputDirectory, "-DpackageName=grafeas", fmt.Sprintf("-DpackageVersion=", ServerVersion)}
```

Which would yield something like:

```shell
java -jar ./swagger-codegen-cli.jar generate -i ./grafeas.swagger.json -l go -o 0.1.0 -c ./config.go.json -DpackageName=grafeas -DpackageVersion=0.1.0
```

Reply:
Got it, thanks so much for the thorough clarifications on this! I agree that more automation-friendly approach would be better, and I like *Version 2/3* because this allows for more automated generation of values. Moving `packageName` out of the configuration seems reasonable. For `packageVersion`, I'm open to automating this as well, with one detail: the intention here is to separate server release version from the client, eg server is at version `0.1.4` and the client is `0.2.1` because of some new features and bug fixes added to the client, which are independent of the server releases. What'd you suggest as the best way to address this, while minimizing the use of `config.go.json`?

Reply:
Ah, okay that makes sense. I was operating under the assumption that there was some coordination of the client version with the server version. We could pretty easily include a `ClientVersion` in the code which is accessible via ldflags when doing generation which could be supplied as a flag or otherwise passed as an environment variable (`CLIENT_VERSION` or `GRAFEAS_CLIENT_VERSION` etc) which would be ingested and would determine the output directory name and the `packageVersion` flag?

I was trying to avoid needing to vendor any additional (external) dependencies so, I don't have anything doing semantic version parsing or incrementing, but the logic _could_ be something like

```go
if len(ClientVersion) == 0 {
  // I've used something like https://github.com/Masterminds/semver in the past
  // for some of the parsing and incrementing functionality
  ClientVersion = autoIncrementPatch()
}
```

But this would rely on version being known at which point you would need a `VERSION`, `version.go`, or I suppose the `config.go.json` file unless you do some logic to extrapolate tag based on current commit (assuming that this repo would be subject to tags and releases).

All of that to say, I guess this could be manually supplied for now via the `config.go.json`, via a ldflag and/or environment variable?

Reply:
Yes, I agree that specifying the client version via `config.go.json`, via ldflag and env var would be great here, as you suggested above. My concern with auto-incrementing versions is that it'd be harder to encode whether micro, minor or major version should change based on the commits, but this is a decision humans can make more effectively.

Happy to merge this once the client version change is in. It sounds like we're on the same page about everything else!

Reply:
Great! I will try to set aside a little bit of time in the next day or two to get that sorted and committed.